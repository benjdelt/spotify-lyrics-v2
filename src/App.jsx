import { Suspense, useEffect, useState } from 'react';

import { getUser, getLyrics, getNowPlaying, getInitialSong, getTrackHistory } from './api';
import { errorTypes, emptyUser, emptySong, defaultLoading } from './utils';

import Layout from './components/Layout';
import Error from './components/Error';
import Nav from './components/Nav';
import Profile from './components/Profile';
import Song from './components/Song';
import Disclaimer from './components/Disclaimer';
import Loader from './components/Loader';


function App() {

  const [user, setUser] = useState(emptyUser);
  const [song, setSong] = useState(emptySong);
  const [trackHistory, setTrackHistory] = useState([]);
  const [loading, setLoading] = useState(defaultLoading);
  const [errorType, setErrorType] = useState("");

  useEffect(() => {
    setLoading({user: false, song: false});
    async function setInitialDatat() {
      setLoading({user: true, song: true});
      const userData = await getUser();
      setUser(prevUser => ({
        ...prevUser,
        name: userData.name,
        avatarUrl: userData.avatar,
      }));
      setLoading(prevLoading => ({
        ...prevLoading,
        user: false,
      }));
      if (userData.name) {
        const songData = await getInitialSong();
        setSong(prevSong => ({
          ...prevSong, 
          title: songData.title,
          artist: songData.artist,
          album: songData.album,
          coverUrl: songData.coverUrl,
          lyrics: songData.lyrics,
        }));
        if (!songData.lyrics.length) {
          setErrorType(errorTypes.noLyricsFound);
        }
        setLoading(prevLoading => ({
          ...prevLoading,
          history: true,
        }));
        const trackHistory = await getTrackHistory();
        setTrackHistory(trackHistory);
        setLoading(prevLoading => ({
          ...prevLoading,
          history: false,
        }));
      }
      setLoading(prevLoading => ({
        ...prevLoading,
        song: false,
      }));     
    }
    setInitialDatat();
  }, []);

  const setToCurrentlyPlaying = async () => {
    setLoading({...loading, song: true});
    setErrorType("");
    const nowPlaying = await getNowPlaying();
    if (nowPlaying.title) {
      setSong(nowPlaying);
      if (!nowPlaying.lyrics.length) {
        setErrorType(errorTypes.noLyricsFound);
      }
    } else {
      setErrorType(errorTypes.noCurrentlyPlaying);
    }
    setLoading({...loading, song: false});
  }

  const setToHistoryTrack = async track => {
    setLoading({...loading, song: true});
    setErrorType("");
    const lyrics = await getLyrics(track.artist, track.title);
    console.log(lyrics)
    setSong({
      title: track.title,
      artist: track.artist,
      album: track.album,
      coverUrl: track.coverUrl,
      lyrics: lyrics,
    })
    if (!lyrics.length) {
      setErrorType(errorTypes.noLyricsFound);
    }
    setLoading({...loading, song: false});
  }

  const logout = () => {
    window.location.hash = "";
    setUser(emptyUser);
    setSong(emptySong);
    setTrackHistory([]);
    setLoading({
      user: false,
      song: false,
      history: false,
    })
  }

  return (
    <Layout 
      nav={<Nav
        loggedIn={!!user.name}
        setToCurrentlyPlaying={setToCurrentlyPlaying}
        trackHistory={trackHistory}
        setToHistoryTrack={setToHistoryTrack}
        loadingHistory={loading.history} 
      />}
    >
      <>
        <Profile user={user} loading={loading.user} logout={logout} />
        {loading.song ? (
          <Loader />
        ) : (
          !user.name ? (
            <Disclaimer />
          ) : ( 
            errorType === errorTypes.noCurrentlyPlaying ? (
              <Error errorType={errorType} />
            ) : (
              <Song song={song} errorType={errorType} />
          )))}
      </>
    </Layout>
  )
}

function LoadingPage() {
  return (
    <Layout nav="" >
      <Loader />
    </Layout>
  )
}

export default function Page() {
  return (
    <Suspense fallback={<LoadingPage />}>
      <App />
    </Suspense>
  )
}
