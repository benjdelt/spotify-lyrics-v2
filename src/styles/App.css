
li {
  list-style: none;
}

button {
  color: #b3b3b3;
  background-color: transparent;
  text-align: left;
  padding: 0;
  border-color: transparent;
  border-width: 0;
  border-style: unset;
  font: unset;
  cursor: pointer;
}

button:focus, a:focus {
  outline: unset;
}

a {
  color: #b3b3b3;
}

/* Media Queries */

@media (max-width: 860px) {
  .menu-text {
    font-size: .75rem;
  }
}

@media (max-width: 732px) {
  .menu-text {
    display: none;
  }
}
