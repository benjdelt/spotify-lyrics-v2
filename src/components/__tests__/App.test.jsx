import { render, cleanup, waitFor } from '@testing-library/react';
import App from '../App';

afterEach(cleanup)

it('should take a snapshot', async () => {
  await waitFor(() => {
    const { asFragment } = render(<App />);
    expect(asFragment(<App />)).toMatchSnapshot();
  })
});
