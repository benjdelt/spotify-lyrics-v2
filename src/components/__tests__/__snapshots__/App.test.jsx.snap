// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should take a snapshot 1`] = `
<DocumentFragment>
  <header
    class="app-header"
  >
    <div
      class="logo"
    >
      <h1>
        <img
          alt="logo"
          src="inverted-logo.png"
        />
        <span
          class="menu-text"
        >
          Spotify Lyrics
        </span>
      </h1>
    </div>
    <nav>
      <ul>
        <li>
          <a
            href="https://github.com/benjdelt/spotify-lyrics-v2"
            rel="noopener noreferrer"
            target="_blank"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-github fa-w-16 "
              data-icon="github"
              data-prefix="fab"
              focusable="false"
              role="img"
              viewBox="0 0 496 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"
                fill="currentColor"
              />
            </svg>
             
            <span
              class="menu-text"
            >
              nav.seeOnGithub
            </span>
          </a>
        </li>
        <li
          class="language"
        >
          <div
            class="dropdown-parent"
          >
            <button
              class="dropdown-label"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-globe-europe fa-w-16 "
                data-icon="globe-europe"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 496 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M248 8C111 8 0 119 0 256s111 248 248 248 248-111 248-248S385 8 248 8zm200 248c0 22.5-3.9 44.2-10.8 64.4h-20.3c-4.3 0-8.4-1.7-11.4-4.8l-32-32.6c-4.5-4.6-4.5-12.1.1-16.7l12.5-12.5v-8.7c0-3-1.2-5.9-3.3-8l-9.4-9.4c-2.1-2.1-5-3.3-8-3.3h-16c-6.2 0-11.3-5.1-11.3-11.3 0-3 1.2-5.9 3.3-8l9.4-9.4c2.1-2.1 5-3.3 8-3.3h32c6.2 0 11.3-5.1 11.3-11.3v-9.4c0-6.2-5.1-11.3-11.3-11.3h-36.7c-8.8 0-16 7.2-16 16v4.5c0 6.9-4.4 13-10.9 15.2l-31.6 10.5c-3.3 1.1-5.5 4.1-5.5 7.6v2.2c0 4.4-3.6 8-8 8h-16c-4.4 0-8-3.6-8-8s-3.6-8-8-8H247c-3 0-5.8 1.7-7.2 4.4l-9.4 18.7c-2.7 5.4-8.2 8.8-14.3 8.8H194c-8.8 0-16-7.2-16-16V199c0-4.2 1.7-8.3 4.7-11.3l20.1-20.1c4.6-4.6 7.2-10.9 7.2-17.5 0-3.4 2.2-6.5 5.5-7.6l40-13.3c1.7-.6 3.2-1.5 4.4-2.7l26.8-26.8c2.1-2.1 3.3-5 3.3-8 0-6.2-5.1-11.3-11.3-11.3H258l-16 16v8c0 4.4-3.6 8-8 8h-16c-4.4 0-8-3.6-8-8v-20c0-2.5 1.2-4.9 3.2-6.4l28.9-21.7c1.9-.1 3.8-.3 5.7-.3C358.3 56 448 145.7 448 256zM130.1 149.1c0-3 1.2-5.9 3.3-8l25.4-25.4c2.1-2.1 5-3.3 8-3.3 6.2 0 11.3 5.1 11.3 11.3v16c0 3-1.2 5.9-3.3 8l-9.4 9.4c-2.1 2.1-5 3.3-8 3.3h-16c-6.2 0-11.3-5.1-11.3-11.3zm128 306.4v-7.1c0-8.8-7.2-16-16-16h-20.2c-10.8 0-26.7-5.3-35.4-11.8l-22.2-16.7c-11.5-8.6-18.2-22.1-18.2-36.4v-23.9c0-16 8.4-30.8 22.1-39l42.9-25.7c7.1-4.2 15.2-6.5 23.4-6.5h31.2c10.9 0 21.4 3.9 29.6 10.9l43.2 37.1h18.3c8.5 0 16.6 3.4 22.6 9.4l17.3 17.3c3.4 3.4 8.1 5.3 12.9 5.3H423c-32.4 58.9-93.8 99.5-164.9 103.1z"
                  fill="currentColor"
                />
              </svg>
               
              <span
                class="menu-text"
              >
                nav.language
              </span>
                
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-caret-up fa-w-10 "
                data-icon="caret-up"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 320 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M288.662 352H31.338c-17.818 0-26.741-21.543-14.142-34.142l128.662-128.662c7.81-7.81 20.474-7.81 28.284 0l128.662 128.662c12.6 12.599 3.676 34.142-14.142 34.142z"
                  fill="currentColor"
                />
              </svg>
            </button>
            <div
              class="dropdown-menu"
              style="visibility: hidden; position: relative; bottom: 6.5em;"
            >
              <div
                class="dropdown-option"
              >
                <span
                  class="dropdown-option-text"
                >
                  English
                </span>
              </div>
              <div
                class="dropdown-option"
              >
                <span
                  class="dropdown-option-text"
                >
                  Français
                </span>
              </div>
            </div>
          </div>
        </li>
      </ul>
    </nav>
  </header>
  <main
    class="app-main"
  >
    <section
      class="profile"
    >
      <section
        class="loading"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-spinner fa-w-16 fa-pulse "
          data-icon="spinner"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M304 48c0 26.51-21.49 48-48 48s-48-21.49-48-48 21.49-48 48-48 48 21.49 48 48zm-48 368c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zm208-208c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zM96 256c0-26.51-21.49-48-48-48S0 229.49 0 256s21.49 48 48 48 48-21.49 48-48zm12.922 99.078c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48c0-26.509-21.491-48-48-48zm294.156 0c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48c0-26.509-21.49-48-48-48zM108.922 60.922c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.491-48-48-48z"
            fill="currentColor"
          />
        </svg>
      </section>
    </section>
    <section
      class="loading"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-spinner fa-w-16 fa-pulse "
        data-icon="spinner"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M304 48c0 26.51-21.49 48-48 48s-48-21.49-48-48 21.49-48 48-48 48 21.49 48 48zm-48 368c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zm208-208c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zM96 256c0-26.51-21.49-48-48-48S0 229.49 0 256s21.49 48 48 48 48-21.49 48-48zm12.922 99.078c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48c0-26.509-21.491-48-48-48zm294.156 0c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48c0-26.509-21.49-48-48-48zM108.922 60.922c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.491-48-48-48z"
          fill="currentColor"
        />
      </svg>
    </section>
  </main>
</DocumentFragment>
`;
